### DISCUSSION TITLE
Nativewind not working with NX expo library

### DISCUSSION THREAD

**Discussion by MuttakinHasib** [nativewind/nativewind#542] [Q&A] [answered]
Hi @marklawlor, I have a NX monorepo with expo. I have configured the nativewind with the expo application and it's working as expected.
But I have created an expo ui library in this monorepo. for this when I use the className on my ui library the nativewind is not working on this place. I am attaching the project structure.

![image](https://github.com/marklawlor/nativewind/assets/44552983/b6ddc54b-2209-4018-af6d-de1b05343874)

`apps/mobile/babel.config.js`
```js
module.exports = function (api) {
  api.cache(true);
  return {
    presets: ['babel-preset-expo'],
    plugins: [
      // Required for expo-router
      'expo-router/babel',
      'nativewind/babel',
    ],
  };
};
```

`apps/mobile/metro.config.js`
```js
const { withNxMetro } = require('@nx/expo');
const { getDefaultConfig } = require('@expo/metro-config');
const { mergeConfig } = require('metro-config');
const exclusionList = require('metro-config/src/defaults/exclusionList');

const defaultConfig = getDefaultConfig(__dirname, {
  isCSSEnabled: true,
});

const { assetExts, sourceExts } = defaultConfig.resolver;

/**
 * Metro configuration
 * https://facebook.github.io/metro/docs/configuration
 *
 * @type {import('metro-config').MetroConfig}
 */
const customConfig = {
  transformer: {
    babelTransformerPath: require.resolve('react-native-svg-transformer'),
  },
  resolver: {
    assetExts: assetExts.filter((ext) => ext !== 'svg'),
    sourceExts: [...sourceExts, 'svg'],
    blockList: exclusionList([/^(?!.*node_modules).*\/dist\/.*/]),
    unstable_enableSymlinks: true,
    unstable_enablePackageExports: true,
  },
};

module.exports = withNxMetro(mergeConfig(defaultConfig, customConfig), {
  // Change this to true to see debugging info.
  // Useful if you have issues resolving modules
  debug: false,
  // all the file extensions used for imports other than 'ts', 'tsx', 'js', 'jsx', 'json'
  extensions: [],
  // Specify folders to watch, in addition to Nx defaults (workspace libraries and node_modules)
  watchFolders: [],
});

```

`apps/mobile/tailwind.config.js`
```js
/** @type {import('tailwindcss').Config} */
module.exports = {
  content: ['./src/**/*.{js,jsx,ts,tsx}', 'libs/ui/src/**/*.{js,jsx,ts,tsx}'],
  theme: {
    extend: {
      fontFamily: {
        'inter-light': 'Inter_300Light',
        'inter-regular': 'Inter_400Regular',
        'inter-medium': 'Inter_500Medium',
        'inter-semibold': 'Inter_600SemiBold',
        'inter-bold': 'Inter_700Bold',
        'inter-black': 'Inter_900Black',
      },
    },
  },
  plugins: [],
};

```

**Comment by marklawlor** [comment]
NativeWind is very hard to troubleshoot with screenshots/code snippets, and I cannot troubleshoot a project that I cannot clone. I would recommend creating a minimal NX repo and using it to troubleshoot your issue.

**Comment by Sergirsanchez** [comment]
Is there any update on this issue? 🙏🏿

**Comment by eternal-eager-beaver** [comment]
Any progress on this issue ?

**Comment by Unn4m3DD** [comment]
I was having the same issue with an nx project as well, I got it to work with this code inside tailwind.config.js
```js
// tailwind.config.js
module.exports = {
  content: [
    "src/**/*.{js,jsx,ts,tsx}", // pattern for my application code
    "../../libs/my-lib-name/src/**/*.{js,jsx,ts,tsx}", // pattern for the library code
  ],
  theme: { },
  plugins: [ ],
};
```
I've used the command `npx tailwind -o output.css` mentioned here https://www.nativewind.dev/guides/troubleshooting#verify-your-configuration together with a `<View className="bg-blue-500 />` on my library code to debug it, maybe it will help you as well

**Comment by kirstilynn** [comment]
You can use `createGlobPatternsForDependencies`

In your tailwind.config.js:
```
const { createGlobPatternsForDependencies } = require('@nx/react/tailwind');
const { join } = require('path');

module.exports = {
  content: [
    join(
      __dirname,
      '{src,pages,components,app}/**/*!(*.stories|*.spec).{ts,tsx,html}',
    ),
    ...createGlobPatternsForDependencies(__dirname),
  ],
...etc.
```

**Reply by stefan-schweiger** [reply]
Be aware that this might not work when doing production builds via eas (or nx build) as the nx project graph won't exist during build time in that case.

**Reply by kirstilynn** [reply]
> Be aware that this might not work when doing production builds via eas (or nx build) as the nx project graph won't exist during build time in that case.

Yeah I ran into that problem and ended up resolving it by adding logic that runs the command to generate the graph as part of my pre build step to ensure the graph is generated when it builds. I posted my solution on this thread

https://github.com/nativewind/nativewind/issues/1132#issuecomment-2603000504

**Comment by zackdotcomputer** [comment]
Just adding a datapoint that NX+Expo+Expo-Router+NativeWind is working for me locally after the following:
1. Created expo app with `npx nx g @nx/expo:app` (I named mine "mobile-app", so replace that in the sample commands here if you're following along)
2. Switch over to using expo-router following the instructions for a [Manual Installation](https://docs.expo.dev/router/installation/#quick-start) on Expo's site (being sure to use `nx install mobile-app` instead of `npx expo install`)
3. Install NativeWind and dependencies with `npx nx install mobile-app --packages=nativewind,tailwindcss,react-native-reanimated,react-native-safe-area-context`
4. Add or update these files in the app's root folder (not the NX monorepo's root, and not the app's src directory):
**tailwind.config.js:**
```js
/** @type {import('tailwindcss').Config} */
module.exports = {
  content: ["./src/**/*.{js,jsx,ts,tsx}"],
  presets: [require("nativewind/preset")],

  theme: {
    extend: {}
  },
  plugins: []
};
```

**global.css:**
```css
@tailwind base;
@tailwind components;
@tailwind utilities;
```

**nativewind-env.d.ts:**
```typescript
/// <reference types="nativewind/types" />
```

**babel.config.js:**
```js
module.exports = function (api) {
  api.cache(true);
  return {
    presets: [["babel-preset-expo", { jsxImportSource: "nativewind" }], "nativewind/babel"],
    plugins: ["react-native-reanimated/plugin"]
  };
};
```

**metro.config.js**
```js
const { withNxMetro } = require("@nx/expo");
const { getDefaultConfig } = require("@expo/metro-config");
const { mergeConfig } = require("metro-config");
const { withNativeWind } = require("nativewind/metro");

const defaultConfig = getDefaultConfig(__dirname);
const { assetExts, sourceExts } = defaultConfig.resolver;

/**
 * Metro configuration
 * https://reactnative.dev/docs/metro
 *
 * @type {import('metro-config').MetroConfig}
 */
const customConfig = {
  cacheVersion: "mobile",
  transformer: {
    babelTransformerPath: require.resolve("react-native-svg-transformer")
  },
  resolver: {
    assetExts: assetExts.filter((ext) => ext !== "svg"),
    sourceExts: [...sourceExts, "cjs", "mjs", "svg"]
  }
};

async function createConfig() {
  return await withNxMetro(mergeConfig(defaultConfig, customConfig), {
    // Change this to true to see debugging info.
    // Useful if you have issues resolving modules
    debug: false,
    // all the file extensions used for imports other than 'ts', 'tsx', 'js', 'jsx', 'json'
    extensions: [],
    // Specify folders to watch, in addition to Nx defaults (workspace libraries and node_modules)
    watchFolders: []
  }).then((nxConfig) =>
    withNativeWind(nxConfig, {
      input: "./global.css"
    })
  );
}

module.exports = createConfig();
```

4. And finally, add `import "../../global.css";` to the top of `./src/app/_layout.tsx`.

That all took a bit of finagling (and TBH it'd be great if SOMEONE documented all those tweaks needed for NX on an official site) but at the end it appears to be working after a hard cache clear.

**Reply by stefan-schweiger** [reply]
I pretty much followed the same approach independently and it worked fine for me. But now I have started to produce actual "production" builds instead of devclient builds and suddenly my styles in libs are not working anymore. If I place something in the app itself it works without a problem. Just wondered if you ran into a similar issue and if you did any special configuration in the libs?

**Reply by stefan-schweiger** [reply]
Oh shit I figured it out, I was using `createGlobPatternsForDependencies` trying to be smart, but this does not work with eas build as NX is not really in the play anymore during the build...

**Comment by danstepanov** [comment]
https://www.nativewind.dev/docs/guides/using-with-monorepos

**Comment by rawr-code** [comment, marked as answer]
### ✅ [FIX] NativeWind + Nx Monorepo + EAS Build: missing styles from shared UI packages

After struggling with this issue for days — where **NativeWind styles were not applied in EAS builds** (only working in local dev) — I finally found the root cause and a reliable fix.

#### 💥 The Problem
During the EAS build process, Nx doesn’t have a cached ProjectGraph available.  
This breaks `createGlobPatternsForDependencies()` from `@nx/react/tailwind`, causing it to return an **empty array**:

[readCachedProjectGraph] ERROR: No cached ProjectGraph is available.
WARNING: There was an error creating glob patterns, returning an empty array


When that happens, Tailwind never sees your `packages/ui` or `packages/core` components — so **no styles** get bundled.

---

#### 🧩 The Fix (works with EAS builds)
The idea is to manually generate the Nx ProjectGraph before the Metro bundler runs.

In your Expo app’s `package.json` (for example `apps/expo-app/package.json`),  
add the following post-install hook:

```json
"scripts": {
  "eas-build-post-install": "cd ../../ && node tools/scripts/eas-build-post-install.mjs . apps/expo-app"
}
```


Then create the script file:
tools/scripts/eas-build-post-install.mjs


```js
/**
 * EAS Build Post-Install Fix for Nx + NativeWind
 * ------------------------------------------------
 * 1️⃣ Generates the Nx ProjectGraph so createGlobPatternsForDependencies() works.
 * 2️⃣ Ensures node_modules symlink between app and workspace.
 */

import { symlinkSync, existsSync } from "fs";
import { join } from "path";
import { createProjectGraphAsync } from "@nx/devkit";

const [workspaceRoot, projectRoot] = process.argv.slice(2);

(async () => {
  console.log("🧩 Running eas-build-post-install: generating Nx ProjectGraph...");
  try {
    await createProjectGraphAsync();
    console.log("✅ ProjectGraph generated successfully");
  } catch (err) {
    console.error("⚠️ Failed to generate ProjectGraph:", err.message);
  }

  const wsModules = join(workspaceRoot, "node_modules");
  const prModules = join(projectRoot, "node_modules");

  try {
    if (existsSync(wsModules)) {
      console.log("🔗 Symlink already exists, skipping");
    } else {
      symlinkSync(prModules, wsModules, "dir");
      console.log("🔗 Symlink created");
    }
  } catch (err) {
    console.error("❌ Symlink creation failed:", err.message);
  }

  console.log("🏁 Post-install script finished");
  process.exit(0);
})();

```

Make sure @nx/devkit is installed (either at the workspace root or in your app).

🧪 Result

When you re-run your EAS build, you’ll now see something like this in the logs:

```bash
🧩 Running eas-build-post-install: generating Nx ProjectGraph...
✅ ProjectGraph generated successfully
🔗 Symlink already exists, skipping
Generating Tailwind config for app "expo-app"
Content paths: [
  .../apps/expo-app/src/**,
  .../packages/ui/**,
  .../packages/core/**
]

```

Your shared UI components will now have their styles correctly applied in production 🎉

💡 Why this works

createGlobPatternsForDependencies() relies on the Nx ProjectGraph to know which libraries belong to the app.
EAS’s isolated build environment doesn’t persist that cache — so generating it on demand fixes the dependency resolution for Tailwind.

Hopefully this helps anyone stuck with NativeWind + Nx builds where styles disappear in production.
I’ll open a PR to add this to the official [NativeWind Monorepo Guide](https://www.nativewind.dev/docs/guides/using-with-monorepos?utm_source=chatgpt.com)
 so it’s easier to find.
 
 

<img width="1401" height="683" alt="Screenshot 2025-10-05 at 4 22 01 PM" src="https://github.com/user-attachments/assets/f4b89b0f-a18c-44e9-8772-2233a6b39313" />

**Reply by danstepanov** [reply]
Thanks for providing this <3
